include/swift/AST/Decl.h

```diff
@@ -940,9 +940,27 @@ class alignas(1 << DeclAlignInBits) Decl {
 
   bool isPrivateStdlibDecl(bool treatNonBuiltinProtocolsAsPublic = true) const;
 
-  /// Whether this declaration is weak-imported.
-  bool isWeakImported(ModuleDecl *fromModule,
-                      AvailabilityContext fromContext) const;
+  AvailabilityContext getAvailabilityForLinkage() const;
+
+  /// Whether this declaration or one of it's outer contexts has the
+  /// @_weakLinked attribute.
+  bool isWeakImported() const;
+
+  /// Whether this declaration is weak-imported from the given module,
+  /// either because of the presence of the @_weakLinked attribute, or
+  /// because of availability.
+  ///
+  /// If \p fromModule is non-null, always returns false if the
+  /// declaration is part of \p fromModule, or if the declaration has
+  /// availability but it is at least as available as the deployment
+  /// target.
```

Comment:
This "deployment target" part is questionable, since it isn't actually related to `fromModule`. In practice, the only module that will be passed here is the one being compiled, but if you *did* pass another one it could give a technically-incorrect answer.

Reply:
I clarified the comment to explain that fromModule should either be null or the main module. I agree this is not an ideal API either, but it still seems better than alternatives I could think of.